File: CSC330/poker/Julia/HandSorter.jl
include("HandIdentifier.jl")
include("Card.jl")


function sort_hands!(hands)
    sort_hands_by_type!(hands)
    sort_ties!(hands)
end

function sort_hands_by_type!(hands)
    sort!(hands, by=h -> get_type(h), rev=true)  # Fixed: get_type should be called as a function
end

function sort_ties!(hands)
    start_index = 1
    last_type = get_type(hands[1])  # Fixed: get_type should be called as a function

    for i in 1:length(hands)
        hand = hands[i]
        if get_type(hand) != last_type  # Fixed: get_type should be called as a function
            hands[start_index:i-1] = sort_subarray(hands[start_index:i-1])
            start_index = i
            last_type = get_type(hand)  # Fixed: get_type should be called as a function
        elseif i == length(hands)
            hands[start_index:i] = sort_subarray(hands[start_index:i])
        end
    end
end

function sort_subarray(hands)
    hand_type = get_type(hands[1])  # Fixed: get_type should be called as a function

    if hand_type == 9
        return sort!(hands, by=h -> compare_royal_flush(h, hands[1]), rev=true)
    elseif hand_type == 8
        return sort!(hands, by=h -> compare_straight_flush(h, hands[1]), rev=true)
    elseif hand_type == 7
        return sort!(hands, by=h -> compare_four_of_a_kind(h, hands[1]), rev=true)
    elseif hand_type == 6
        return sort!(hands, by=h -> compare_full_house(h, hands[1]), rev=true)
    elseif hand_type == 5
        return sort!(hands, by=h -> compare_flush(h, hands[1]), rev=true)
    elseif hand_type == 4
        return sort!(hands, by=h -> compare_straight(h, hands[1]), rev=true)
    elseif hand_type == 3
        return sort!(hands, by=h -> compare_three_of_a_kind(h, hands[1]), rev=true)
    elseif hand_type == 2
        return sort!(hands, by=h -> compare_two_pair(h, hands[1]), rev=true)
    elseif hand_type == 1
        return sort!(hands, by=h -> compare_pair(h, hands[1]), rev=true)
    else
        return sort!(hands, by=h -> compare_high_card(h, hands[1]), rev=true)
    end
end

function compare_royal_flush(h1, h2)
    return get_suit(get_sorted_cards(h1)[1]) < get_suit(get_sorted_cards(h2)[1]) ? -1 : 1  # Fixed: get_suit and get_sorted_cards should be called as functions
end

function compare_straight(h1, h2)
    l1, l2 = get_sorted_cards(h1), get_sorted_cards(h2)
    highest_card_comparison = compare_highest_card(l1, l2)
    if highest_card_comparison != 0
        return highest_card_comparison
    end

    suit1 = is_royal_straight(l1) ? get_suit(l1[1]) : get_suit(l1[end])  # Fixed: get_suit should be called as a function
    suit2 = is_royal_straight(l2) ? get_suit(l2[1]) : get_suit(l2[end])  # Fixed: get_suit should be called as a function
    return suit1 < suit2 ? -1 : 1
end

function compare_straight_flush(h1, h2)
    l1, l2 = get_sorted_cards(h1), get_sorted_cards(h2)
    highest_card_comparison = compare_highest_card(l1, l2)
    if highest_card_comparison != 0
        return highest_card_comparison
    end
    return get_suit(l1[1]) < get_suit(l2[1]) ? -1 : 1  # Fixed: get_suit should be called as a function
end

function compare_four_of_a_kind(h1, h2)
    c1 = get_cards_occuring_n_times(get_sorted_cards(h1), 4)[1]
    c2 = get_cards_occuring_n_times(get_sorted_cards(h2), 4)[1]
    return compare_card_values(c1, c2)
end

function compare_full_house(h1, h2)
    c1 = get_cards_occuring_n_times(get_sorted_cards(h1), 3)[1]
    c2 = get_cards_occuring_n_times(get_sorted_cards(h2), 3)[1]
    return compare_card_values(c1, c2)
end

function compare_flush(h1, h2)
    l1, l2 = get_sorted_cards(h1), get_sorted_cards(h2)
    highest_card_comparison = compare_highest_card(l1, l2)
    return highest_card_comparison != 0 ? highest_card_comparison : (get_suit(l1[1]) < get_suit(l2[1]) ? -1 : 1)  # Fixed: get_suit should be called as a function
end

function compare_three_of_a_kind(h1, h2)
    c1 = get_cards_occuring_n_times(get_sorted_cards(h1), 3)[1]
    c2 = get_cards_occuring_n_times(get_sorted_cards(h2), 3)[1]
    return compare_card_values(c1, c2)
end

function compare_two_pair(h1, h2)
    pairs1 = get_cards_occuring_n_times(get_sorted_cards(h1), 2)
    pairs2 = get_cards_occuring_n_times(get_sorted_cards(h2), 2)
    highest_card_comparison = compare_highest_card(pairs1, pairs2)
    return highest_card_comparison != 0 ? highest_card_comparison : compare_highest_card(get_cards_occuring_n_times(get_sorted_cards(h1), 1), get_cards_occuring_n_times(get_sorted_cards(h2), 1))
end

function compare_pair(h1, h2)
    pairs1 = get_cards_occuring_n_times(get_sorted_cards(h1), 2)
    pairs2 = get_cards_occuring_n_times(get_sorted_cards(h2), 2)
    return compare_highest_card(pairs1, pairs2)
end

function compare_high_card(h1, h2)
    l1, l2 = get_sorted_cards(h1), get_sorted_cards(h2)
    return compare_highest_card(l1, l2)
end

function compare_highest_card(l1, l2)
    value_list1 = map(c -> get_value(c), l1)
    value_list2 = map(c -> get_value(c), l2)
    sort!(value_list1, rev=true)
    sort!(value_list2, rev=true)

    for (v1, v2) in zip(value_list1, value_list2)
        if v1 == 0 || v2 == 0
	    if v1 == 1
		return 1
	    elseif v2 == 1
		return -1
	    end
	end    
	if v1 < v2
            return -1
        elseif v1 > v2
            return 1
        end
    end
    return 0
end

function get_cards_occuring_n_times(card_list, n)
    return [card for card in unique(card_list) if count(x -> x == card, card_list) == n]
end

File: CSC330/poker/Go/HandSorter.go
package main

import "sort"

func SortHands(hands []*Hand) {
    SortHandsByType(hands)
    SortTies(hands)
}

func SortHandsByType(hands []*Hand) {
    sort.Slice(hands, func(i, j int) bool {
        return hands[i].GetType() > hands[j].GetType()
    })
}

// SortTies splits hands into subarrays of similar types and sorts them
func SortTies(hands []*Hand) {
    var sortedHands []*Hand
    startIndex := 0
    lastType := hands[0].GetType()

    for i, hand := range hands {
        if hand.GetType() != lastType || i == len(hands)-1 {
            if hand.GetType() != lastType {
                sortedHands = append(sortedHands, SortSubarray(hands[startIndex:i])...)
                startIndex = i
            } else {
                sortedHands = append(sortedHands, SortSubarray(hands[startIndex:])...)
            }
            lastType = hand.GetType()
        }
    }
    copy(hands, sortedHands)
}

func SortSubarray(hands []*Hand) []*Hand {
    handType := hands[0].GetType()

    switch handType {
    case 9: // Royal Flush
        sort.Slice(hands, func(i, j int) bool {
            return compareRoyalFlush(hands[i], hands[j]) < 0
        })
    case 8: // Straight Flush
        sort.Slice(hands, func(i, j int) bool {
            return compareStraightFlush(hands[i], hands[j]) < 0
        })
    case 7: // Four of a Kind
        sort.Slice(hands, func(i, j int) bool {
            return compareFourOfAKind(hands[i], hands[j]) < 0
        })
    case 6: // Full House
        sort.Slice(hands, func(i, j int) bool {
            return compareFullHouse(hands[i], hands[j]) < 0
        })
    case 5: // Flush
        sort.Slice(hands, func(i, j int) bool {
            return compareFlush(hands[i], hands[j]) < 0
        })
    case 4: // Straight
        sort.Slice(hands, func(i, j int) bool {
            return compareStraight(hands[i], hands[j]) < 0
        })
    case 3: // Three of a Kind
        sort.Slice(hands, func(i, j int) bool {
            return compareThreeOfAKind(hands[i], hands[j]) < 0
        })
    case 2: // Two Pair
        sort.Slice(hands, func(i, j int) bool {
            return compareTwoPair(hands[i], hands[j]) < 0
        })
    case 1: // Pair
        sort.Slice(hands, func(i, j int) bool {
            return comparePair(hands[i], hands[j]) < 0
        })
    default: // High Card
        sort.Slice(hands, func(i, j int) bool {
            return compareHighCard(hands[i], hands[j]) < 0
        })
    }

    return hands
}

// Compare Functions
func compareRoyalFlush(h1, h2 *Hand) int {
    return compareHighestCard(h1.GetSortedCards(), h2.GetSortedCards())
}

func compareStraightFlush(h1, h2 *Hand) int {
    return compareHighestCard(h1.GetSortedCards(), h2.GetSortedCards())
}

func compareFourOfAKind(h1, h2 *Hand) int {
    c1 := getCardsOccurringNTimes(h1.GetSortedCards(), 4)[0]
    c2 := getCardsOccurringNTimes(h2.GetSortedCards(), 4)[0]
    return compareHighestCard(c1, c2)
}

func compareFullHouse(h1, h2 *Hand) int {
    triplet1 := getCardsOccurringNTimes(h1.GetSortedCards(), 3)[0]
    triplet2 := getCardsOccurringNTimes(h2.GetSortedCards(), 3)[0]
    pair1 := getCardsOccurringNTimes(h1.GetSortedCards(), 2)[0]
    pair2 := getCardsOccurringNTimes(h2.GetSortedCards(), 2)[0]

    // Compare triplets first
    result := compareHighestCard(triplet1, triplet2)
    if result != 0 {
        return result
    }
    // If triplets are equal, compare pairs
    return compareHighestCard(pair1, pair2)
}

func compareFlush(h1, h2 *Hand) int {
    return compareHighestCard(h1.GetSortedCards(), h2.GetSortedCards())
}

func compareStraight(h1, h2 *Hand) int {
    return compareHighestCard(h1.GetSortedCards(), h2.GetSortedCards())
}

func compareThreeOfAKind(h1, h2 *Hand) int {
    return compareHighestCard(getCardsOccurringNTimes(h1.GetSortedCards(), 3)[0], getCardsOccurringNTimes(h2.GetSortedCards(), 3)[0])
}

func compareTwoPair(h1, h2 *Hand) int {
    pairs1 := getCardsOccurringNTimes(h1.GetSortedCards(), 2)
    pairs2 := getCardsOccurringNTimes(h2.GetSortedCards(), 2)

    // Compare highest pair
    result := compareHighestCard(pairs1[0], pairs2[0])
    if result != 0 {
        return result
    }

    // If highest pairs are equal, compare second pair
    return compareHighestCard(pairs1[1], pairs2[1])
}

func comparePair(h1, h2 *Hand) int {
    return compareHighestCard(getCardsOccurringNTimes(h1.GetSortedCards(), 2)[0], getCardsOccurringNTimes(h2.GetSortedCards(), 2)[0])
}

func compareHighCard(h1, h2 *Hand) int {
    return compareHighestCard(h1.GetSortedCards(), h2.GetSortedCards())
}

// Helper Functions
func compareHighestCard(hand1, hand2 []Card) int {
    for i := 0; i < len(hand1); i++ {
        if hand1[i].rank > hand2[i].rank {
            return -1
        } else if hand1[i].rank < hand2[i].rank {
            return 1
        }
        if hand1[i].suit > hand2[i].suit {
            return -1
        } else if hand1[i].suit < hand2[i].suit {
            return 1
        }
    }
    return 0
}

func getCardsOccurringNTimes(cardList []Card, n int) [][]Card {
    rankMap := make(map[int][]Card)
    for _, card := range cardList {
        rankMap[card.rank] = append(rankMap[card.rank], card)
    }
    var result [][]Card
    for _, cards := range rankMap {
        if len(cards) == n {
            result = append(result, cards)
        }
    }
    return result
}

func adjustValue(card Card) int {
    switch card.suit {
    case 1: // Spades
        return card.rank + 10
    case 2: // Hearts
        return card.rank + 5
    case 3: // Clubs
        return card.rank + 2
    default: // Diamonds
        return card.rank
    }
}

